Improve Reset emulation

mpilgrem opened the pull request:
Currently, the emulation (for versions of Windows without an ANSI-aware console) fills blank space with 'colour 0' (usually black) and 'resets' to 'colour  7 on colour 0' (usually silver on black). This may upset PowerShell users (or others) who have a different default - for example, `Reset` does not return them to where they started.

This proposal changes the emulation so that the default colours are those of the console when the emulation starts.

The default colours are captured in module `System.Console.ANSI.Windows.Detect` and preserved: `isANSIEnabled` is now a value of type `ANSIEnabledState` (rather than `Bool`) which has a constructor `NotANSIEnabled ConsoleDefaultState`. The emulated version of the affected functions (eg `hClearScreen` and `hSetSGR`) are then a function of the 'console default state' (as well as the other arguments).

This sees: (1) some of the original `Common-Include.hs` split out into `Common-Include-Enabled.hs` or, as applicable, `Common-Include-Emulator.hs`; and (2) `if then else` choices in module  `System.Console.ANSI.Windows` become `case of` choices.

mpilgrem commented:
I have tested this improved emulation on Windows 7 and Windows 10 (with ANSI disabled).

feuerbach commented:
Thanks a lot, Mike. Your PR descriptions are throrough; could you also:

1. copy them into the git commit message and
2. incorporate the relevant parts of it as source code comments?

This way if someone studies the code later, they won't have to dig up the pull request to understand the context.

This applies to both this PR and #28.

Otherwise, this looks good to me. You are doing excellent work, and I am quite happy for you to commit directly to the repo; I'll grant you the rights. Of course, feel free to open PRs if you want to solicit feedback.

mpilgrem commented:
@feuerbach, I am still a git novice, and uncertain what commiting directly to the repo entails. I don't want to break anything. Does it entail just clicking on the green 'Merge pull request' button when 'All checks have passed' and the 'branch has no conflicts with the base branch'?

I have extended the commit message, added some documentation and added a couple of helper functions to module `System.Console.ANSI.Windows` (`nativeOrEmulated` and `nativeOrEmulatedWithDefault`) to cut back on the repetitive `case of` code in that module.

feuerbach commented:
The way I prefer to do it is just `git push` from the command line. I assume you already use `git push` to push to your repo. Now you can add this repo as another remote and push to it directly when you are ready.

But you can certainly use github interface, too. In that case, please select "Rebase and merge" from that green dropdown, so that github doesn't create unnecessary merge commits.
